### PR TITLE
feat(zoe): caretaker-style revocable, ownable

### PR DESCRIPTION
**Pull request by erights**
closes: #XXXX
refs: #8020 #8517 #2070

## Description

Alternative to #8020 and #8517

Unlike #8020 provides a caretaker-style revoker, which is a much better base for building ownables

Unlike #8517 provides ownables that are much easier to layer on top of existing implementations.

The basic idea of an "ownable object" is to provide general and lower level support for approximately the pattern that vaults implement manually in an adhoc manner: A *ownable object* is an object with a `makeTransferInvitation` method that revokes the object itself, in the sense of inactivating it — it no longer provides any authority. Instead, the exclusively transferable portion of its authority is encapsulated by the returned invitation, in an allegedly legible manner. That invitation can be redeemed, via an `offer` , for a new ownable object in which the transferable authority from the original ownable object is now exercisable.

We call this an "ownable object" since rights of ownership include
- the right to exercise (use)
- the right to exclusively transfer the rights of ownership
- the right to exclude others from being able to exercise and transfer
- the ability to legibly offer the exclusive transfer of ownership

Objects are about rights to exercise (use) and share, but not exclusive transfer, and not legibly.

ERTP is about legible exclusive transfer but not exercise.

The lifecycle of an ownable object, transmuting between an ephemeral ownable object and an invitation to get a new “equivalent” ownable object, combines all attributes together. But the attributes are never directly present at the same time. That’s why it is the lifecycle that combines these.

This notion of "ownable" and "ownership rights" has strong echoes in programming language notions of "ownership types", "linear types", "affine types", and "reference capabilities", in languages ranging from Rust to Pony to Move, enable mutable objects that can be exercised (use) and exclusive transfer, but not legibility.

However, none of these, as far as we are aware, has the two-phase lifecycle of our ownable objects. Thus, they do they provide for legibility of offers to transfer, which is crucial for secure exchange and offer safety. To produce a legible description of the transferable right (which we include in the invitation `customDetails`), we need to simultaneously stabilize the object, denying any further exercise until the invitation is redeemed. Otherwise, the "seller" could seem to be selling something whose value is invalidated by the seller's own further use.

See [*Necessity* specifications for robustness](https://dl.acm.org/doi/abs/10.1145/3563317)

See https://groups.google.com/g/cap-talk/c/9OU9ZNoB3os/m/gXeu2zs4AgAJ?utm_medium=email&utm_source=footer for a useful real-world analogy for the two-phase ownership cycle.

### Security Considerations

Must ensure that revocation really denies all further authority from the original ownable object, even from messages that are already in the air but not yet delivered.

Need to verify that the legibility can meet invitation legibility goals.

### Scaling Considerations

The kind of intrinsic revocability of #8020 could provide scaling benefits of dropping bookkeeping for used-up objects such as payments. However, we are not currently pursuing that revocation plan. It is conceivable that we could do something similar with the caretaker-like revocability of this PR, but we do not currently plan to do so.

### Documentation Considerations

Once ready to be used, not only should this be documented, we should be able to simplify several other APIs in ways that should make them more explainable.

### Testing Considerations

First tests are included.

### Upgrade Considerations

Given that these are used with the durable zone, as intended:
- The transfer invitations have durable handlers and so are themselves durable and should survive upgrade.
- The ownable objects themselves are defined to be durable, so they can survive upgrade if their underlying object can.

### CODE REVIEW

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [inline packages/zoe/src/contractSupport/prepare-ownable.js]: Remember to change this from `zcf` to a `makeInvitation` function. This is both more POLA, and is more flexible. For example, the escrow transfer of ownable ZCFSeat could be expressed as a virtualized `makeInvitation`

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (Chris-Hibbert) [inline packages/zoe/src/contractSupport/prepare-ownable.js]: I still think customDetails doesn't say enough. Do you mind `inviteDetails`? Do you have a different theory?

File: packages/zoe/src/contracts/ownable-counter.js
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/ownable-counter.js:1]: I don't think this is an interesting sample contract. It shows how ownables work, but isn't a useful stand-alone contract. Having a version in a test in `contractSupport` seems more apropos.

File: packages/vat-data/test/test-prepare-revocable.js
Comment (dckc) [inline packages/vat-data/test/test-prepare-revocable.js]: x is optional, right?

```suggestion
    /** @param {number} [x] */
```

File: packages/vat-data/test/test-prepare-revocable.js
Comment (dckc) [inline packages/vat-data/test/test-prepare-revocable.js]: When teaching our stuff, I hope to skip the `prepareExoClass` layer and go right to `zone`.
I can imagine you don't want this test to depend on anything that it doesn't have to, but... can folks who only want to use `zone` use `prepareRevokableKit`?

File: packages/vat-data/test/test-prepare-revocable.js
Comment (dckc) [inline packages/vat-data/test/test-prepare-revocable.js]: I'm always leery of committing to error messages in our APIs. I tend to do something like...

```suggestion
    message: /revoked/,
```

File: packages/zoe/src/typeGuards.js
Comment (dckc) [inline packages/zoe/src/typeGuards.js]: that could be a separate fix

**Review by dckc — COMMENTED** [review]
I don't see any important problems.

I defer to @Chris-Hibbert to approve.

File: packages/zoe/src/typeGuards.js
Comment (erights) [reply packages/zoe/src/typeGuards.js]: https://github.com/Agoric/agoric-sdk/pull/8748

File: packages/vat-data/test/test-prepare-revocable.js
Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: Done

File: packages/zoe/src/typeGuards.js
Comment (erights) [reply packages/zoe/src/typeGuards.js]: Done

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: Done

File: packages/vat-data/test/test-prepare-revocable.js
Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: I am conflicted on this point. On one hand, any test of the contents of an error message is a fragile golden, in that the only real requirement is that the diagnostic is helpful, informative, and not misleading. We want to be able to continue to improve error messages on all three of these dimensions, and golden tests put friction in the way. I have experienced this myself, and the friction is non-trivial.

OTOH, the mechanisms creating these supposedly-helpful error messages can be complicated, and the error messages can accidentally become less helpful or more misleading over time without anyone otherwise noticing. So my practice has been towards more complete more golden errors. 

I don't have a good resolution to this tension. Perhaps that practice should change in general. But either way, I'd rather not deviate from my general practice just here.

So, resolving without change. If you disagree, please unresolve this conversation.

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: I agree with this sentiment. But
- `inviteDetails` is not grammatical because `invite` is a verb, suggesting `invitationDetails`
- But `invitationDetails` already has an existing meaning in our system. See https://github.com/Agoric/agoric-sdk/blob/87e4ef498bd57ac1577e46ecdcc3bb96c0a98c88/packages/zoe/src/zoeService/invitationQueries.js#L8 and https://github.com/Agoric/agoric-sdk/blob/87e4ef498bd57ac1577e46ecdcc3bb96c0a98c88/packages/zoe/src/zoeService/types.js#L41 . 
- The invitation's `customDetails` is a subrecord of an invitation's invitation-details record.

In light of all this, I went with `getInvitationCustomDetails`. I normally hate names this long, but since it needs to be understood as distinct from the object's "normal" method names, the length isn't terrible.

Done.

File: packages/zoe/src/contracts/ownable-counter.js
Comment (erights) [reply packages/zoe/src/contracts/ownable-counter.js:1]: Done. Moved to test/unitTests/contracts/ , which is where its tester is.

File: packages/vat-data/test/test-prepare-revocable.js
Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: This is an excellent point. I agree. But it requires a deeper refactoring, moving `prepareRevocable` to a package that can depend on zones. (It is currently in `vat-data` which cannot.)

Putting this PR into Draft until I do that refactor.

Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: @michaelfig , since I'm about to zone-ify, I added you as reviewer. Thanks.

Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: Once prepare-revocable.js is refactored to depend on the base-zone package, and given that we will want to migrate it to endo when we migrate base-zone to endo, where should it be moved now? https://github.com/Agoric/agoric-sdk/pull/8749 

Ok to move prepare-revocable.js to base-zone itself?

A separate question is what should it be named once it is refactored to depend on zones, and therefore not necessarily limited to the "prepare-for-durable" level of abstraction?

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: A dissenting voice on whether "invite" can be used as a noun: https://grammarist.com/usage/invite/

Nevertheless, I'm gonna stick with `getInvitationCustomDetails` because it appears on the ownable alongside `makeTransferInvitation`, and I want to suggest that these are related.

File: packages/vat-data/test/test-prepare-revocable.js
Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: See https://github.com/endojs/endo/pull/1965
https://github.com/Agoric/agoric-sdk/pull/8750

File: packages/vat-data/test/test-prepare-revocable.js
Comment (michaelfig) [reply packages/vat-data/test/test-prepare-revocable.js]: I expect that the `@agoric/base-zone` (or `@endo/zone`) package should only contain the bare minimum API (objects and collections) that it currently has.  Could `prepare-revocable.js` be part of the contract utilities that `@agoric/zoe` currently exports, and moved only if/when we split `@agoric/zoe` into multiple packages?

Another option is to follow the trail blazed by the `@agoric/whenable` package I built in #8742.  I had originally considered putting Whenable functionality into `@agoric/base-zone`, but thought better of it and resisted the temptation.

Comment (michaelfig) [reply packages/vat-data/test/test-prepare-revocable.js]: Oh, I misunderstood.

> moving `prepareRevocable` to a package that can depend on zones.

`prepareRevocable` can stay in `@agoric/vat-data` and take a `"devDependencies"` dependency on `@agoric/base-zone` (later `@endo/base-zone`), without any circular dependencies.  Neither of those packages currently depend on the other one.

Comment (michaelfig) [reply packages/vat-data/test/test-prepare-revocable.js]: Ah, I think I've come around to putting it in `base-zone`.  I think it would fit well as a `@property {...} revocableKit` on the `@typedef {object} ExoZone` definition in `packages/base-zone/src/types.js`.

Sound good?

Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: > When teaching our stuff, I hope to skip the `prepareExoClass` layer and go right to `zone`.
> I can imagine you don't want this test to depend on anything that it doesn't have to, but... can folks who only want to use `zone` use `prepareRevokableKit`?

Zonified! Done.

Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: > A separate question is what should it be named once it is refactored to depend on zones, and therefore not necessarily limited to the "prepare-for-durable" level of abstraction?

Not yet renamed.

> Ah, I think I've come around to putting it in base-zone. I think it would fit well as a @property {...} revocableKit on the @typedef {object} ExoZone definition in packages/base-zone/src/types.js.

For now, I've moved prepare-revocable.js to zoe instead alongside prepare-ownable.js . I suggest we leave it there until base-zone itself moves to endo, at which time we can migrate prepare-revocable as well.

Comment (erights) [reply packages/vat-data/test/test-prepare-revocable.js]: I think this is done enough for the points raised in this conversation, so resolving. Please unresolve and comment further if you disagree.

**Comment by erights** [comment]
With the zonification done, this is again R4R. Not urgent, but a lot of future development rests on this, so I'd like to see it make progress. Thanks!

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (dckc) [inline packages/zoe/src/contractSupport/prepare-ownable.js]: prettier goes a little nutso with vertical space. Consider:

```suggestion
        const { underlying } = this.state;
        const { revoker } = this.facets;
```

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (dckc) [inline packages/zoe/src/contractSupport/prepare-ownable.js]: again, to avoid wasting vertical space, consider:

```suggestion
        const { underlying } = this.state;
```

File: packages/zoe/test/unitTests/contractSupport/test-prepare-revocable.js
Comment (dckc) [inline packages/zoe/test/unitTests/contractSupport/test-prepare-revocable.js]: multiple arguments of the same type next to each other means there's little to no help from the machine in getting them in the right order.

Having 2 un-named optional args at the end also suggests that more may come later. And it means that to specify the last one, you have to put `undefined` in the 2nd-to-last spot.

Maybe put all the args but `zone` in an options object?

File: packages/zoe/test/unitTests/contracts/test-ownable-counter.js
Comment (dckc) [inline packages/zoe/test/unitTests/contracts/test-ownable-counter.js:55]: we rely on the ownable object to supply the custom details? not zoe?

Comment (dckc) [reply packages/zoe/test/unitTests/contracts/test-ownable-counter.js:55]: I think I see now... after `makeTransferInvitation`, we can ask zoe.

**Review by dckc — COMMENTED** [review]
looks pretty good.

Before I approve... is this expected to be part of the **developer API product**?
If so, I would try to generate docs and see what they look like.

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: Done

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: Done

File: packages/zoe/test/unitTests/contractSupport/test-prepare-revocable.js
Comment (erights) [reply packages/zoe/test/unitTests/contractSupport/test-prepare-revocable.js]: I moved only the optional parameters into the options object, which is my normal habit for options objects. If there are too many remaining required parameters, I might still try to make these non-positional somehow. But in this case, there were few enough required parameters left that I kept them positional.

PTAL. I'm resolving under the theory that this is done enough. If you disagree please unresolve and comment. Thanks.

File: packages/zoe/test/unitTests/contracts/test-ownable-counter.js
Comment (erights) [reply packages/zoe/test/unitTests/contracts/test-ownable-counter.js:55]: Yup. Resolving.

**Comment by erights** [comment]
> looks pretty good.
> 
> Before I approve... is this expected to be part of the **developer API product**? If so, I would try to generate docs and see what they look like.

@dckc 
I've never done that before, but I'd love to! Where do I start?

**Comment by dckc** [comment]
For adding generated docs, an example to follow is

 - #8720

After you add stuff like that, run `yarn typedoc` from the top `agoric-sdk` directory. I think the results go in an `api-doc` directory.

**Comment by erights** [comment]
> For adding generated docs, an example to follow is
> 
> * [more typedoc #8720](https://github.com/Agoric/agoric-sdk/pull/8720)
> 
> After you add stuff like that, run `yarn typedoc` from the top `agoric-sdk` directory. I think the results go in an `api-doc` directory.

[api-docs.zip](https://github.com/Agoric/agoric-sdk/files/14133296/api-docs.zip)

At api-docs/modules/_agoric_zoe.src_contractSupport_prepare_ownable.html
and api-docs/modules/_agoric_zoe.src_contractSupport_prepare_revocable.html

I like it!

**Comment by erights** [comment]
@dckc 
I don't think I did that right. New commit coming up

**Comment by erights** [comment]
[api-docs.zip](https://github.com/Agoric/agoric-sdk/files/14133545/api-docs.zip)


Done. Now they are just in the zoe/contractSupport module. They get kinda lost in there, bit altogether it seems like the right thing to do. Opinions?

**Comment by dckc** [comment]
> ... Now they are just in the zoe/contractSupport module. They get kinda lost in there, but altogether it seems like the right thing to do. Opinions?

Yeah... I think we're at the "better than nothing" stage of developing reference docs.
I don't have any better sense of what this stuff should look like.

**Comment by erights** [comment]
@dckc , do you have any remaining concerns?

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (michaelfig) [inline packages/zoe/src/contractSupport/prepare-ownable.js]: These errors can be avoided by typing of `prepareRevocableKit`.  I'll sketch a way to do it below.

File: packages/zoe/src/contractSupport/prepare-revocable.js
Comment (michaelfig) [inline packages/zoe/src/contractSupport/prepare-revocable.js]:
```js
/**
 * @template [U=any]
 * @typedef {object} RevocableKitThis
 * @property {RevocableKit<U>} facets
 * @property {{ underlying: U }} state
 */
```

```suggestion
 *   (this: RevocableKitThis<U>, ...args: any[]) => any
```

**Review by michaelfig — APPROVED** [review]

**Review by dckc — APPROVED** [review]
I had some concerns about how this integrates with upcoming releases (cf. #8578) but it doesn't impact any existing on-chain bundles, and that seems to address these concerns to @mhofman 's satisfaction, which makes it fine by me.

File: packages/zoe/src/contractSupport/prepare-ownable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-ownable.js]: Done

File: packages/zoe/src/contractSupport/prepare-revocable.js
Comment (erights) [reply packages/zoe/src/contractSupport/prepare-revocable.js]: Much better, thanks. Done

**Pull request merged**
